File: C#/Leetcode/BinaryTree/PathSumII.cs
using LeetcodeSolutions.DataStructures;
using System.Collections.Generic;

namespace LeetcodeSolutions.BinaryTree
{
    // Leetcode 113
    // Submission Detail : https://leetcode.com/problems/path-sum-ii/discuss/36695/Java-Solution:-iterative-and-recursive
    public class PathSumII
    {
        public IList<IList<int>> PathSum(BinaryTreeNode root, int sum)
        {
            IList<IList<int>> result = new List<IList<int>>();
            IList<int> list = new List<int>();

            Helper(root, result, list, sum);
            return result;
        }

        private void Helper(BinaryTreeNode node, IList<IList<int>> result, IList<int> list, int sum)
        {
            list.Add(node.Val);

            if (node.Left == null && node.Right == null && node.Val == sum)
            {
                result.Add(new List<int>(list));
                return;
            }

            if (node.Left != null)
            {
                Helper(node.Left, result, list, sum - node.Val);
                list.RemoveAt(list.Count - 1);
            }

            if (node.Right != null)
            {
                Helper(node.Right, result, list, sum - node.Val);
                list.RemoveAt(list.Count - 1);
            }
        }
    }
}

File: Java/DFS/Tree/PathSumII.java
package DFS.Tree;

import DataStructures.TreeNode;

import java.util.ArrayList;
import java.util.List;

public class PathSumII {
  // Leetcode 113:
  // Submission Detail: https://leetcode.com/problems/path-sum-ii/submissions/1651190904

  List<List<Integer>> result;
  public PathSumII() {
    result = new ArrayList<>();
  }

  // Tx = O(n)
  // Sx = O(n + k log n) {k: number of paths}
  public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
    dfs(root, new ArrayList<>(), targetSum);

    return result;
  }

  private void dfs(TreeNode node, List<Integer> pathNodes, int remainingTargetSum) {
    // Base case: Do nothing for an empty node.
    if(node == null)
      return;

    pathNodes.add(node.val);

    // Base Case: Left node. Check if the path equals targetsum and add the resultant array to the list.
    if(node.left == null && node.right == null) {
      if(node.val == remainingTargetSum) {
        // Create a new arraylist copy so that we don't modify it after adding.
        result.add(new ArrayList<>(pathNodes));
        pathNodes.remove(pathNodes.size()-1);

        return;
      }
    }

    dfs(node.left, pathNodes, remainingTargetSum-node.val);
    dfs(node.right, pathNodes, remainingTargetSum-node.val);

    pathNodes.remove(pathNodes.size()-1);
  }
}
